flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkBusyJoinReorderRule.java

```diff
@@ -0,0 +1,688 @@
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.sql.SqlExplainLevel;
+import org.apache.calcite.tools.RelBuilder;
+import org.apache.calcite.tools.RelBuilderFactory;
+import org.apache.calcite.util.ImmutableBitSet;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Flink busy join reorder rule, which will convert {@link MultiJoin} to a busy join tree.
+ *
+ * <p>In this busy join reorder strategy, we will first try to reorder all the inner join type
+ * inputs in the multiJoin, and then add all outer join type inputs on the top.
+ *
+ * <p>First, reordering all the inner join type inputs in the multiJoin. We adopt the concept of
+ * level in dynamic programming, and the latter layer will use the results stored in the previous
+ * layers. First, we put all inputs (each input in {@link MultiJoin}) into level 0, then we build
+ * all two-inputs join at level 1 based on the {@link FlinkCost} of level 0, then we will build
+ * three-inputs join based on the previous two levels, then four-inputs joins ... etc, util we
+ * reorder all the inner join type inputs in the multiJoin. When building m-inputs join, we only
+ * keep the best plan (have the lowest {@link FlinkCost}) for the same set of m inputs. E.g., for
+ * three-inputs join, we keep only the best plan for inputs {A, B, C} among plans (A J B) J C, (A J
+ * C) J B, (B J C) J A.
+ *
+ * <p>Second, we will add all outer join type inputs in the MultiJoin on the top.
+ */
+public class FlinkBusyJoinReorderRule extends RelRule<FlinkBusyJoinReorderRule.Config>
+        implements TransformationRule {
+
+    public static final LoptOptimizeJoinRule MULTI_JOIN_OPTIMIZE =
+            LoptOptimizeJoinRule.Config.DEFAULT.toRule();
+
+    /** Creates an SparkJoinReorderRule. */
+    protected FlinkBusyJoinReorderRule(Config config) {
+        super(config);
+    }
+
+    @Deprecated // to be removed before 2.0
+    public FlinkBusyJoinReorderRule(RelBuilderFactory relBuilderFactory) {
+        this(Config.DEFAULT.withRelBuilderFactory(relBuilderFactory).as(Config.class));
+    }
+
+    @Deprecated // to be removed before 2.0
+    public FlinkBusyJoinReorderRule(
+            RelFactories.JoinFactory joinFactory,
+            RelFactories.ProjectFactory projectFactory,
+            RelFactories.FilterFactory filterFactory) {
+        this(RelBuilder.proto(joinFactory, projectFactory, filterFactory));
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
```

Comment:
If's better we can introduce a `FlinkJoinReorderRule`, which could choose the specific join reorder algorithm based on  `dpThreshold `

Reply:
> If's better we can introduce a `FlinkJoinReorderRule`, which could choose the specific join reorder algorithm based on `dpThreshold `
Done！